### PR TITLE
[GLUTEN-4796][VL] Force fallback for orc char type scan

### PR DESCRIPTION
**Pull request by kerwin-zk**
## What changes were proposed in this pull request?
issue #4796 

The issue resolved in #4259  dealt with the exception that occurred when a char column was included in the select statement. This issue fixes the exception that occurs when the char column is not included in the select statement.

## How was this patch tested?
unit tests

### CODE REVIEW

**Comment by jinchengchenghh** [comment]
Can you add a test to track this?

**Comment by kerwin-zk** [comment]
> Can you add a test to track this?

@jinchengchenghh done

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/GlutenSQLQuerySuite.scala [hunk 1]
```diff
@@ -112,4 +113,17 @@ class GlutenSQLQuerySuite extends SQLQuerySuite with GlutenSQLTestsTrait {
...
+      sql("CREATE TABLE hive_orc(name CHAR(10), id INT) STORED AS ORC")
+      sql("INSERT INTO hive_orc VALUES('a', 1)")
+    }
+    withSQLConf(HiveUtils.CONVERT_METASTORE_ORC.key -> "true") {
+      withTable("hive_orc") {
+        checkAnswer(sql("SELECT name, id FROM hive_orc"), Row("a", 1))
+        checkAnswer(sql("SELECT id FROM hive_orc"), Row(1))
+      }
+    }
+  }
```
Comment (zhouyuan) [inline gluten-ut/spark33/src/test/scala/org/apache/spark/sql/GlutenSQLQuerySuite.scala:128]: should also apply to spark 32/34 ?

**Review by zhouyuan — COMMENTED** [review]
Thanks for helping on this. The CI failures seems not related.
Could you please help to add unit tests for spark32/34 also?

**Comment by kerwin-zk** [comment]
@zhouyuan cc

**Review by zhouyuan — APPROVED** [review]
👍 
CC @rui-mo

File: gluten-core/src/main/scala/io/glutenproject/execution/BasicScanExecTransformer.scala [hunk 2]
```diff
@@ -88,12 +89,29 @@ trait BasicScanExecTransformer extends LeafTransformSupport with BaseDataSource
   }
 
   override protected def doValidateInternal(): ValidationResult = {
+    def addStringFields(
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/BasicScanExecTransformer.scala:92]: nit: addStringFields -> appendStringFields, and maybe extract this function out to generate all the fields needed for validation.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 3]
```diff
@@ -151,9 +155,11 @@ object BackendSettings extends BackendSettingsApi {
                 if mapType.valueType.isInstanceOf[ArrayType] =>
               "ArrayType as Value in MapType"
             case StructField(_, stringType: StringType, _, metadata)
-                if CharVarcharUtils
-                  .getRawTypeString(metadata)
-                  .getOrElse(stringType.catalogString) != stringType.catalogString =>
+                if GlutenConfig.getConf.forceOrcCharTypeScanFallbackEnabled && CHAR_TYPE_PATTERN
+                  .findFirstIn(CharVarcharUtils
+                    .getRawTypeString(metadata)
+                    .getOrElse(stringType.catalogString))
```
Comment (rui-mo) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:161]: Maybe extract as a new function `isCharType`.

**Review by rui-mo — COMMENTED** [review]
Thanks.

**Comment by kerwin-zk** [comment]
> Thanks.

@rui-mo done

File: gluten-core/src/main/scala/io/glutenproject/execution/BasicScanExecTransformer.scala [hunk 4]
```diff
@@ -88,12 +89,18 @@ trait BasicScanExecTransformer extends LeafTransformSupport with BaseDataSource
...
 
   override protected def doValidateInternal(): ValidationResult = {
+    var fields = schema.fields
+
+    this match {
+      case transformer: FileSourceScanExecTransformer =>
+        fields = appendStringFields(transformer.relation.schema, fields)
+      case transformer: HiveTableScanExecTransformer =>
+        fields = appendStringFields(transformer.getDataSchema, fields)
+      case _ =>
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/BasicScanExecTransformer.scala:99]: Do we need BatchScanExec here?

Comment (kerwin-zk) [reply gluten-core/src/main/scala/io/glutenproject/execution/BasicScanExecTransformer.scala:101]: @rui-mo  I haven't encountered it here, but in theory, as long as there is a native scan of orc, it is needed. I have added in the code.

**Review by rui-mo — APPROVED** [review]

**Review by rui-mo — APPROVED** [review]

**Pull request merged**